### PR TITLE
task3_business_logic

### PR DESCRIPTION
**Pull request by egor-geller**

### CODE REVIEW

File: model/src/main/java/com/epam/esm/entity/Order.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    @Column(name = "purchase_date", updatable = false, columnDefinition = "TIMESTAMP() default CURRENT_TIMESTAMP() on update CURRENT_TIMESTAMP()")
+    @CreationTimestamp
+    private ZonedDateTime purchaseDate;
+
+    @ManyToOne
+    @JoinColumn(name = "id_user", referencedColumnName = "id")
+    private User user;
+
+    @ManyToMany
+    @JoinTable(
```
Comment (nickaden) [inline model/src/main/java/com/epam/esm/entity/Order.java:51]: If we've update certifcate's price, we will see discrepancy in order cost and certificate price sum. Could we save certificate price at the moment of order creation?

File: model/src/main/java/com/epam/esm/audit/AuditEntity.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package com.epam.esm.audit;
```
Comment (nickaden) [inline model/src/main/java/com/epam/esm/audit/AuditEntity.java:1]: Why do you need it?

File: repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Autowired
+    public UserRepositoryImpl(EntityManager entityManager, PaginationContext paginationContext) {
+        this.paginationContext = paginationContext;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public List<User> findAll() {
+        return entityManager.createQuery(FIND_ALL_USERS, User.class)
```
Comment (nickaden) [inline repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java:31]: Can we move `findAll` method to abstract repository? The only things changed are class and query

Comment (nickaden) [reply repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java:31]: You could also extract all another methods of repository

File: repository/src/main/java/com/epam/esm/repository/repositoryinterfaces/Repository.java [hunk 4]
```diff
@@ -16,15 +18,15 @@
      * Fina a tag or certificate by its unique Id
      *
      * @param id id of a tag or certificate
-     * @return {@link Optional} of {@link Tag} or {@link Certificate} entity.
+     * @return {@link Optional} of {@link Tag}, {@link Certificate}, {@link Order} or {@link User} entity.
```
Comment (nickaden) [inline repository/src/main/java/com/epam/esm/repository/repositoryinterfaces/Repository.java]: You don't know ALL implementation of your interface. You don't need to list of possible return values here

File: repository/src/main/java/com/epam/esm/ErrorType.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.epam.esm;
+
+public enum ErrorType {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/ErrorType.java]: Are you going to put error codes here for other exceptions? The name of the enumeration looks rather abstract

File: repository/src/main/java/com/epam/esm/exception/InvalidPageContextException.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.epam.esm.exception;
+
+import com.epam.esm.ErrorType;
+
+public class InvalidPageContextException extends RuntimeException{
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/exception/InvalidPageContextException.java]: The name looks rather ambiguous. Perhaps you can rename it to Pagination Exception?

File: repository/src/main/java/com/epam/esm/repository/PaginationContext.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    private static final int MAX_PAGE_SIZE = 50;
+    private static final int DEFAULT_PAGE = 1;
+    private static final int DEFAULT_PAGE_SIZE = 10;
+
+    private int page;
+    private int pageSize;
+
+    private PaginationContext(){}
+
+    public PaginationContext of(Integer page, Integer pageSize) {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/PaginationContext.java]: Naming methods.

File: repository/src/main/java/com/epam/esm/repository/PaginationContext.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        this.page = page;
+        this.pageSize = pageSize;
+
+        if (page < MIN_PAGE) {
+            throw new InvalidPageContextException(INVALID_PAGE_NUMBER, page);
+        }
+
+        if (pageSize < MIN_PAGE_SIZE || pageSize > MAX_PAGE_SIZE) {
+            throw new InvalidPageContextException(INVALID_PAGE_SIZE, pageSize);
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/PaginationContext.java]: I think that the validation logic can be put in a separate place

File: repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.epam.esm.repository.builder;
+
+public final class OrderQueries {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java]: Why is this class in the builder package?

Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java]: Why did you decide to use separate immutable classes to store queries to each entity?

File: repository/src/main/java/com/epam/esm/repository/impl/OrderRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public List<Order> findByUserId(Long id) {
+        return entityManager.createQuery(SELECT_ORDER_BY_USER_ID, Order.class)
+                .setParameter(ID, id)
+                .setFirstResult(paginationContext.getStartPage())
+                .setMaxResults(paginationContext.getLengthOfContext())
+                .getResultList();
+    }
+
+    @Transactional
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/impl/OrderRepositoryImpl.java:68]: Why are you using @Transactional at the DAO level?

File: repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    @Transactional
+    @Override
+    public User create(User user) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Transactional
+    @Override
+    public void delete(User user) {
+        throw new UnsupportedOperationException();
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java]: You are violating the SOLID: Interface segregation principle

File: repository/src/main/java/com/epam/esm/repository/repositoryinterfaces/Repository.java (hunk 4 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/repositoryinterfaces/Repository.java]: fixed

File: repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java (hunk 11 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java]: fixed, but not sure

File: repository/src/main/java/com/epam/esm/repository/impl/OrderRepositoryImpl.java (hunk 10 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/impl/OrderRepositoryImpl.java:68]: As the transaction only occurs in this level, I don't need all the methods in the service or web layers "know" about the transaction operation

File: repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java (hunk 9 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java]: Mainly for future queries

File: repository/src/main/java/com/epam/esm/ErrorType.java (hunk 5 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/ErrorType.java]: fixed to more specific

File: repository/src/main/java/com/epam/esm/exception/InvalidPageContextException.java (hunk 6 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/exception/InvalidPageContextException.java]: fixed

File: repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java (hunk 3 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/impl/UserRepositoryImpl.java:31]: fixed, but not sure

File: repository/src/main/java/com/epam/esm/repository/PaginationContext.java (hunk 7 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/PaginationContext.java]: fixed

File: repository/src/main/java/com/epam/esm/repository/PaginationContext.java (hunk 8 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/PaginationContext.java]: fixed

File: repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java (hunk 9 above)
Comment (egor-geller) [reply repository/src/main/java/com/epam/esm/repository/builder/OrderQueries.java]: fixed

File: model/src/main/java/com/epam/esm/audit/AuditEntity.java (hunk 2 above)
Comment (egor-geller) [reply model/src/main/java/com/epam/esm/audit/AuditEntity.java:1]: I need this to save/track a record in db for AuditListener operations such as creation, updation, deletion of an entity

File: model/src/main/java/com/epam/esm/entity/Order.java (hunk 1 above)
Comment (egor-geller) [reply model/src/main/java/com/epam/esm/entity/Order.java:51]: Should I make another field in Order class which will save "at the moment" certificate price?

**Pull request merged**